File: app/models/space.py
from sqlalchemy import Column, String, Boolean, Integer, TIMESTAMP, text, ForeignKey, CheckConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base


class Space(Base):
    __tablename__ = "spaces"

    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("gen_random_uuid()")
    )

    residence_id = Column(
        UUID(as_uuid=True),
        ForeignKey("residences.id", ondelete="CASCADE"),
        nullable=False
    )

    name = Column(String, nullable=False)
    space_type = Column(String, nullable=False)

    template_type = Column(String, nullable=False, server_default=text("'single_room'"))
    standard = Column(String, nullable=False, server_default=text("'nsfas'"))

    is_rentable = Column(Boolean, server_default=text("false"))
    capacity = Column(Integer, server_default=text("0"))

    floor = Column(Integer, nullable=True)
    notes = Column(String, nullable=True)

    created_at = Column(
        TIMESTAMP(timezone=True),
        server_default=func.now()
    )

    is_active = Column(Boolean, server_default=text("true"))

    updated_at = Column(
        TIMESTAMP(timezone=True),
        server_default=func.now(),
        onupdate=func.now()
    )

    archived_at = Column(TIMESTAMP(timezone=True), nullable=True)

    residence = relationship("Residence", backref="spaces")

    __table_args__ = (
        CheckConstraint(
            "space_type IN ('room','bathroom','kitchen','common','other')",
            name="spaces_type_check"
        ),
    )

File: mobile/lib/models/space.dart
class Space {
  final String id;
  final String residenceId;
  final String name;
  final String spaceType;
  final String templateType;
  final String standard;
  final bool isRentable;
  final int capacity;
  final int? floor;
  final String? notes;
  final bool isActive;

  const Space({
    required this.id,
    required this.residenceId,
    required this.name,
    required this.spaceType,
    required this.templateType,
    required this.standard,
    required this.isRentable,
    required this.capacity,
    required this.floor,
    required this.notes,
    required this.isActive,
  });

  factory Space.fromJson(Map<String, dynamic> json) {
    return Space(
      id: json['id'].toString(),
      residenceId: json['residence_id'].toString(),
      name: json['name']?.toString() ?? 'Unnamed space',
      spaceType: json['space_type']?.toString() ?? '',
      templateType: json['template_type']?.toString() ?? '',
      standard: json['standard']?.toString() ?? '',
      isRentable: json['is_rentable'] == true,
      capacity: _asInt(json['capacity']),
      floor: _asNullableInt(json['floor']),
      notes: json['notes']?.toString(),
      isActive: json['is_active'] != false,
    );
  }

  bool get isRoom {
    return spaceType == 'room';
  }
}

int _asInt(Object? value) {
  if (value is int) return value;
  if (value is num) return value.toInt();

  return int.tryParse(value?.toString() ?? '') ?? 0;
}

int? _asNullableInt(Object? value) {
  if (value == null) return null;

  return _asInt(value);
}
